pkg/gatewayserver/io/ws/ws.go

```diff
@@ -299,217 +267,63 @@ func (s *srv) handleTraffic(c echo.Context) (err error) {
 				if err != nil {
 					logger.WithError(err).Warn("Failed to send ping message")
 					conn.Disconnect(err)
+					s.formatter.Disconnect(ctx, uid)
 					ws.Close()
 					return
 				}
 			case down := <-conn.Down():
 				dlTime := time.Now()
-				scheduledMsg := down.GetScheduled()
 
-				// The first 16 bits of XTime gets the session ID from the upstream latestXTime and the other 48 bits are concentrator timestamp accounted for rollover.
-				sID := atomic.LoadInt32(&sessionID)
-				concentratorTime, ok := conn.TimeFromTimestampTime(scheduledMsg.Timestamp)
+				concentratorTime, ok := conn.TimeFromTimestampTime(down.GetScheduled().Timestamp)
 				if !ok {
 					logger.Warn("No clock synchronization")
 					continue
 				}
-				xTime := int64(sID)<<48 | (int64(concentratorTime) / int64(time.Microsecond) & 0xFFFFFFFFFF)
-				dnmsg := messages.FromDownlinkMessage(ids, down.GetRawPayload(), scheduledMsg, int64(s.tokens.Next(down.CorrelationIDs, dlTime)), dlTime, xTime)
-				msg, err := dnmsg.MarshalJSON()
+				dnmsg, err := s.formatter.FromDownlink(uid, *down, concentratorTime, dlTime)
 				if err != nil {
 					logger.WithError(err).Warn("Failed to marshal downlink message")
 					continue
 				}
 
 				logger.Info("Send downlink message")
 				wsWriteMu.Lock()
-				err = ws.WriteMessage(websocket.TextMessage, msg)
+				err = ws.WriteMessage(websocket.TextMessage, dnmsg)
 				wsWriteMu.Unlock()
 				if err != nil {
 					logger.WithError(err).Warn("Failed to send downlink message")
 					conn.Disconnect(err)
+					s.formatter.Disconnect(ctx, uid)
 					return
 				}
 			}
 		}
 	}()
 
-	recordXTime := func(timestamp int64, server time.Time) {
-		// The session is the 16 MSB.
-		atomic.StoreInt32(&sessionID, int32(timestamp>>48))
-		conn.SyncWithGatewayConcentrator(
-			// The concentrator timestamp is the 32 LSB.
-			uint32(timestamp&0xFFFFFFFF),
-			server,
-			// The Basic Station epoch is the 48 LSB.
-			scheduling.ConcentratorTime(time.Duration(timestamp&0xFFFFFFFFFF)*time.Microsecond),
-		)
-	}
-
 	for {
 		_, data, err := ws.ReadMessage()
 		if err != nil {
 			logger.WithError(err).Debug("Failed to read message")
 			return err
 		}
-
-		typ, err := messages.Type(data)
+		downMsg, err := s.formatter.HandleUp(ctx, data, ids, conn, time.Now())
```

Comment:
Confusing variable name; is it a response message to the uplink message?

That is downstream, but `downMsg` seems a lot like a downlink message.